filters/auth/oauth.go

```diff
@@ -0,0 +1,429 @@
+		cfg        oauth2.Config
+		authClient *authClient
+	}
+
+	filter struct {
+		typ        roleCheckType
+		authClient *authClient
+		realm      string
+		scopes     []string
+		kv         kv
+	}
+	kv map[string]string
+)
+
+var (
+	errInvalidAuthorizationHeader = errors.New("invalid authorization header")
+	errInvalidToken               = errors.New("invalid token")
+)
+
+func getToken(r *http.Request) (string, error) {
+	const b = "Bearer "
+	h := r.Header.Get(authHeaderName)
+	if !strings.HasPrefix(h, b) {
+		return "", errInvalidAuthorizationHeader
+	}
+
+	return h[len(b):], nil
+}
+
+func unauthorized(ctx filters.FilterContext, uname string, reason rejectReason) {
+	ctx.StateBag()[logfilter.AuthUserKey] = uname
+	ctx.StateBag()[logfilter.AuthRejectReasonKey] = string(reason)
+	ctx.Serve(&http.Response{StatusCode: http.StatusUnauthorized})
+}
+
+func authorized(ctx filters.FilterContext, uname string) {
+	ctx.StateBag()[logfilter.AuthUserKey] = uname
+}
+
+func getStrings(args []interface{}) ([]string, error) {
+	s := make([]string, len(args))
+	var ok bool
+	for i, a := range args {
+		s[i], ok = a.(string)
+		if !ok {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	return s, nil
+}
+
+// all checks that all strings in the left are also in the
+// right. Right can be a superset of left.
+func all(left, right []string) bool {
+	if len(left) > len(right) {
+		return false
+	}
+
+	for _, l := range left {
+		for _, r := range right {
+			if l != r {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// intersect checks that one string in the left is also in the right
+func intersect(left []string, right []string) bool {
+	for _, l := range left {
+		for _, r := range right {
+			if l == r {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// jsonGet requests url with Bearer auth header if `auth` was given
+// and writes into doc.
+func jsonGet(url, auth string, doc interface{}) error {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+
+	if auth != "" {
+		req.Header.Set(authHeaderName, "Bearer "+auth)
+	}
+
+	rsp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer rsp.Body.Close()
+	if rsp.StatusCode != 200 {
+		return errInvalidToken
+	}
+
+	d := json.NewDecoder(rsp.Body)
+	return d.Decode(doc)
+}
+
+func (ac *authClient) getTokeninfo(token string) (map[string]interface{}, error) {
+	var a map[string]interface{}
+	err := jsonGet(ac.urlBase, token, &a)
+	return a, err
+}
+
+func newSpec(typ roleCheckType, cfg oauth2.Config) filters.Spec {
+	return &spec{typ: typ, cfg: cfg}
+}
+
+// Options to configure auth providers
+type Options struct {
+	// TokenURL is the tokeninfo URL able to return information
+	// about a token.
+	TokenURL string
+	// AuthType is the type of authnz function you want to
+	// use. Examples are the values "authAnyScope" or "authAllScope",
+	// defined in constants AuthAnyScopeName and AuthAllScopeName.
+	AuthType string
```

Comment:
We should make `roleCheckType` exported, and accept enumerated values in the options instead of strings, because the options are used programmatically.

Another way would to just have separate NewAuthAny, NewAuthAll... constructors. See the header filter as an example: https://github.com/zalando/skipper/blob/master/filters/builtin/headerfilter.go#L72